File: modules/admin/impl/default-ts/seed.ts
/**
 * Seed script — creates the default admin account.
 *
 * Usage:
 *   bun run seed
 *
 * Reads ADMIN_DEFAULT_EMAIL and ADMIN_DEFAULT_PASSWORD from .env,
 * creates the account via Better Auth signup, then sets role to "admin".
 */

const email = process.env.ADMIN_DEFAULT_EMAIL ?? 'admin@localhost'
const password = process.env.ADMIN_DEFAULT_PASSWORD ?? 'changeme'
const authUrl = process.env.BETTER_AUTH_URL ?? 'http://localhost:8080'
const secret = process.env.BETTER_AUTH_SECRET ?? ''

async function seed() {
  console.log(`Creating admin account: ${email}`)

  // 1. Sign up the admin user
  const signupRes = await fetch(`${authUrl}/api/auth/sign-up/email`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name: 'Admin', email, password }),
  })

  if (!signupRes.ok && signupRes.status !== 409) {
    const body = await signupRes.text()
    console.error(`Failed to create admin: ${signupRes.status} ${body}`)
    process.exit(1)
  }

  if (signupRes.status === 409) {
    console.log('Admin account already exists, updating role...')
  }

  // 2. Get the user ID
  const listRes = await fetch(
    `${authUrl}/api/auth/admin/list-users?searchField=email&searchValue=${encodeURIComponent(email)}&limit=1`,
    { headers: { 'x-better-auth-secret': secret } },
  )
  const listData = (await listRes.json()) as { users: Array<{ id: string }> }
  const userId = listData.users?.[0]?.id
  if (!userId) {
    console.error('Could not find admin user after signup')
    process.exit(1)
  }

  // 3. Set role to admin
  const roleRes = await fetch(`${authUrl}/api/auth/admin/set-role`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'x-better-auth-secret': secret,
    },
    body: JSON.stringify({ userId, role: 'admin' }),
  })

  if (!roleRes.ok) {
    console.error(`Failed to set admin role: ${roleRes.status}`)
    process.exit(1)
  }

  console.log(`Admin account ready: ${email} (role: admin)`)
}

seed().catch((err) => {
  console.error(err)
  process.exit(1)
})

File: modules/admin/impl/default-go/seed.go
// Package admin contains the seed command for creating the default admin account.
//
// Build and run:
//
//	go run cmd/seed/main.go
package admin

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net/http"
	"os"
)

// Seed creates the default admin account using Better Auth's REST API.
func Seed() error {
	email := envOr("ADMIN_DEFAULT_EMAIL", "admin@localhost")
	password := envOr("ADMIN_DEFAULT_PASSWORD", "changeme")
	authURL := envOr("BETTER_AUTH_URL", "http://localhost:8080")
	secret := os.Getenv("BETTER_AUTH_SECRET")

	fmt.Printf("Creating admin account: %s\n", email)

	// 1. Sign up the admin user
	signupBody, _ := json.Marshal(map[string]string{
		"name": "Admin", "email": email, "password": password,
	})
	signupResp, err := http.Post(authURL+"/api/auth/sign-up/email", "application/json", bytes.NewReader(signupBody))
	if err != nil {
		return fmt.Errorf("signup request failed: %w", err)
	}
	defer signupResp.Body.Close()

	if signupResp.StatusCode != http.StatusOK && signupResp.StatusCode != http.StatusConflict {
		return fmt.Errorf("signup returned %d", signupResp.StatusCode)
	}
	if signupResp.StatusCode == http.StatusConflict {
		fmt.Println("Admin account already exists, updating role...")
	}

	// 2. Get the user ID
	listURL := fmt.Sprintf("%s/api/auth/admin/list-users?searchField=email&searchValue=%s&limit=1", authURL, email)
	listReq, _ := http.NewRequest(http.MethodGet, listURL, nil)
	listReq.Header.Set("x-better-auth-secret", secret)
	listResp, err := http.DefaultClient.Do(listReq)
	if err != nil {
		return fmt.Errorf("list users request failed: %w", err)
	}
	defer listResp.Body.Close()

	var listData struct {
		Users []struct {
			ID string `json:"id"`
		} `json:"users"`
	}
	if err := json.NewDecoder(listResp.Body).Decode(&listData); err != nil {
		return fmt.Errorf("decode list users: %w", err)
	}
	if len(listData.Users) == 0 {
		return fmt.Errorf("could not find admin user after signup")
	}
	userID := listData.Users[0].ID

	// 3. Set role to admin
	roleBody, _ := json.Marshal(map[string]string{"userId": userID, "role": "admin"})
	roleReq, _ := http.NewRequest(http.MethodPost, authURL+"/api/auth/admin/set-role", bytes.NewReader(roleBody))
	roleReq.Header.Set("Content-Type", "application/json")
	roleReq.Header.Set("x-better-auth-secret", secret)
	roleResp, err := http.DefaultClient.Do(roleReq)
	if err != nil {
		return fmt.Errorf("set role request failed: %w", err)
	}
	defer roleResp.Body.Close()

	if roleResp.StatusCode != http.StatusOK {
		return fmt.Errorf("set role returned %d", roleResp.StatusCode)
	}

	fmt.Printf("Admin account ready: %s (role: admin)\n", email)
	return nil
}

func envOr(key, fallback string) string {
	if v := os.Getenv(key); v != "" {
		return v
	}
	return fallback
}
